Add helper download.InMemory for downloading files without writing to disk

anjannath opened the pull request:
this helper is put in a separate package pkg/download/inmemory to avoid import cycles which occurs due to the network package being imported in the pkg/download package, the cycle is network->ssh->constants->version as a result we'll not be able to import pkg/download in pkg/crc/version

currently the helper is used for downloading release-info.json file and the signed bundle hashes file sha256sum.txt.sig

anjannath committed:
Add helper download.InMemory for downloading files without writing to disk

Currently we have separate code that downloads the release-info.json file
and separate code that downloads the signed bundle hashes file, both  the
files are downloaded in memory this helper will unify that code and we'll
use this helper to download both of these files

It sets the http User-Agent to `crc/<version>` which is currently used for
downloading the file release-info.json

cfergeau commented:
> this helper is put in a separate package pkg/download/inmemory to avoid import cycles which occurs due to the network package being imported in the pkg/download packag

This import cycle already caused issues in the past, I think another way of breaking the cycle is by moving the proxy code (`HTTPTransport()`) to its own package

cfergeau commented on pkg/crc/machine/bundle/metadata.go:
```diff
@@ -299,19 +299,15 @@ func getBundleDownloadInfo(preset crcPreset.Preset) (*download.RemoteFile, error
 		return "", err
 	}
-	if err := res.Body.Close(); err != nil {
+	if err := res.Close(); err != nil {
```
Should be a `defer res.Close()` just before `signedHashes, err := io.ReadAll(res)`, otherwise you won't close the file when `ReadAll` fails.

cfergeau commented on pkg/download/inmemory/inmemory.go:
```diff
@@ -0,0 +1,45 @@
+// Download takes an endpoint in the form of a string (URL) or a *http.Request
+// allowing caller to pass a custom http request to the downloader library, it
+// takes a http.RoundTripper as the second argument to be used by the client
+func Download[T string | *http.Request](endpoint T, transport http.RoundTripper) (io.ReadCloser, error) {
```
You can avoid the use of the generics by always setting the user-agent in this method, which makes sense in my opinion.

anjannath replied:
i think in zabbix we ignore requests with http agent `crc *` from the download data count, with the current use of this (for getting release info and bundle hashes) it should be okay, will remove the type parameters and set the user-agent in the function itself

anjannath replied:
more import cycles, to set the user agent to `crc/<crc_version>`we have to import the `version` package in `inmemory` package and the version package has to import the `inmemory` package to download the `release-info.json`

cfergeau replied:
We still want to be consistent imo and always use the same user agent. I guess for now you could add a `userAgent string` parameter to `Download`, and a `version.UserAgent()`, and pass it explicitly. Otherwise this needs more code movement.

gbraad replied:
does it functionally change anything? what are you planning to do with the 'useragent'; track it as telemetry?

cfergeau replied:
Downloading the crc version file is done with user agent set, downloading the signature file is done without a user agent set, and there are 2 code paths in the download code to deal with both situations.
I don't mind dropping the user agent or always setting it, but it would be better to always do the same thing.

See also https://github.com/crc-org/crc/pull/3693#discussion_r1213949744

anjannath replied:
When we request the `release-info.json` file from the CRC codebase to check if a new version exists, we set the `User-Agent` for that call as, e.g `crc 2.20.0` this was added to make sure that we are not skewing the download counts by our own tool making http requests to the download location.

and in zabbix where this download data is collected filters the requests having user agent `crc *` so these are not added to the download count

anjannath replied:
`User-Agent` seems is not used to filter out http requests from `crc`, not sure how i reached that, but it doesn't make sense also as for collecting the download data from splunk we can use a query language and can filter out the `release-info.json` or the `sha256sum.txt.sig` files.

gbraad commented:
> I think another way of breaking the cycle is by moving the proxy code (HTTPTransport()) to its own package

is this the alternative you prefer or is this something that can ALSO be done.

cfergeau commented:
> > I think another way of breaking the cycle is by moving the proxy code (HTTPTransport()) to its own package
> 
> is this the alternative you prefer or is this something that can ALSO be done.

That's merely an alternative suggestion which can be ignored or done, no strong feeling. However, it's at least the 3rd time we get import cycles because of `HTTPTransport` being in the `network` package (happened to me in some experiments, quite sure Praveen saw that in one of his PRs, but avoided the cycle differently, now Anjan hits it too). So we could try to solve this once and for all.

anjannath committed:
version: use download.InMemory to fetch release-info.json

anjannath committed:
bundle: use download.InMemory to fetch signed bundle hashes

as a result of using the helper the User-Agent for the http request
is `crc/<version>`

this was not the case before, but since we were setting the user-agent
for downloading the `release-info.json` file it makes sense to use the
same User-Agent for downloading the file `sha256sum.txt.sig`

this also fixes #3686

anjannath commented:
i haven't looked at solving this in a better way so not sure how much work it'll be, created https://github.com/crc-org/crc/issues/3697 for planning/tracking

cfergeau committed:
move HTTPTransport and proxy related code to its own package httpproxy

cfergeau reviewed (commented):
Looks good to me, there are several mentions of `inmemory.Download` in the commit logs which no longer are accurate.
You should also mention in the last commit log that this should help with bug xxx (the one with the download timeouts).

anjannath commented:
> Looks good to me, there are several mentions of `inmemory.Download` in the commit logs which no longer are accurate.
> You should also mention in the last commit log that this should help with bug xxx (the one with the download timeouts).

Thanks, updated the commit logs!

anjannath commented:
Changed the function name to `download.InMemory`, as otherwise, gorevive complains about the name.. 

```
Error: pkg/download/download.go:91:6: exported: func name will be used as download.DownloadInMemory by other packages, and that stutters; consider calling this InMemory (revive)
func DownloadInMemory(url, userAgent string, transport http.RoundTripper) (io.ReadCloser, error) {
     ^
```
